### PR TITLE
[codex] Improve generated SEO metadata

### LINKED ISSUE #534: Track full-site SEO metadata pass
> *This was generated by AI during implementation tracking.*

## Summary

Implement a full-site SEO pass focused on search-result clarity, crawl metadata accuracy, and structured-data completeness.

## Change outline

- Add optional `seoTitle` and `seoDescription` fields for content collections so visible article/project copy can stay expressive while search metadata stays concise.
- Shorten generated `<title>` and `<meta name="description">` output for long blog posts and project pages flagged by the generated-site crawl.
- Add RSS discovery via `<link rel="alternate" type="application/rss+xml" href="/rss.xml">`.
- Enrich blog/project collection pages with `ItemList` JSON-LD.
- Expand blog post and project detail JSON-LD with clearer page/entity metadata.
- Replace build-time sitemap timestamps with conservative content-derived `lastmod` dates for blog URLs; omit `lastmod` where the repo has no reliable content date.
- Update SEO specs, agent docs, and generated-output tests to lock the new behavior.

## Validation target

Before opening the PR, verify:

- `npm run lint`
- `npm run typecheck`
- `npm run test`
- `npm run test:e2e`
- generated `dist/` SEO crawl: no missing metadata, broken internal links, duplicate H1s, missing JSON-LD, missing RSS alternate links, or overlong indexed titles/descriptions.

## Tracking

A PR will follow from the implementation branch and should close this issue once merged.

### PR DESCRIPTION
**Pull request by nathanjohnpayne**
Authoring-Agent: codex

Closes #534

## Summary

This PR tightens generated SEO output across the site without changing the visible editorial copy. It adds optional SEO-only content fields, uses them for long blog/project metadata, enriches collection and detail structured data, advertises the RSS feed globally, and makes sitemap `lastmod` values content-derived instead of build-time generated.

## Changes

- Add optional `seoTitle` / `seoDescription` fields to blog content and optional `seoDescription` to project content.
- Use concise blog post search titles/descriptions while preserving full visible H1s and article copy.
- Use concise project detail meta descriptions where project card/hero copy is intentionally longer.
- Add global RSS alternate metadata.
- Add `ItemList` JSON-LD for blog and project collection pages.
- Expand BlogPosting and project JSON-LD with canonical URLs, date/language/free-access metadata, keywords, and repository identity links where applicable.
- Replace sitemap build timestamps with content-derived blog dates and omit `lastmod` where no reliable source date exists.
- Update SEO specs, agent docs, and generated-output regression tests.

## Self-Review

- Verified the generated-site crawl before and after the change. Before: long indexed titles/descriptions and build-time sitemap timestamps. After: no crawl findings for missing metadata, broken internal links, duplicate H1s, missing JSON-LD, missing RSS alternate links, or overlong indexed metadata.
- Confirmed sitemap `lastmod` values now come only from blog frontmatter dates and are omitted for routes without reliable content dates.
- Confirmed visible article/project H1s and descriptions remain unchanged while search metadata uses concise SEO fields.
- Left unrelated untracked `bugs/` notes out of the branch.

## Validation

- `npm run build`
- `npm run lint`
- `npm run typecheck`
- `npm run test`
- `npm run test:e2e`
- Generated `dist/` SEO crawl: 16 pages, 0 issues; sitemap `lastmod` limited to blog index/post URLs.

### CODE REVIEW

**Comment by nathanjohnpayne** [comment]
## External Review Required

**PR:** #535—[codex] Improve generated SEO metadata
**Tracking Issue:** #534
**Branch:** codex/seo-metadata-pass
**Author Agent:** codex

### Summary

This PR improves generated SEO output across the site without changing visible editorial copy. It adds optional SEO-only content fields, uses them for long blog/project metadata, enriches collection/detail JSON-LD, advertises RSS globally, and makes sitemap `lastmod` values content-derived instead of build-time generated.

### Focus Areas

- Verify the sitemap `lastmod` strategy is conservative enough: blog URLs use frontmatter dates; pages without reliable dates omit `lastmod`.
- Check that `seoTitle` / `seoDescription` reduce search-result verbosity without creating metadata/content drift.
- Review JSON-LD changes for correctness, especially blog `BlogPosting`, collection `ItemList`, and project `SoftwareApplication` / keyword / repository identity fields.
- Confirm the generated-output tests assert behavior rather than overfitting to implementation details.

### Observations from Internal Review

- The diff is 299 changed lines (`269 insertions, 30 deletions`), just under the repo's configured 300-line external-review threshold, and it does not touch configured protected paths. This handoff is human-requested for tracking issue #534 rather than automatically threshold-triggered.
- `dateModified` currently equals `datePublished` for blog posts because the blog schema has no explicit updated-date field. That is accurate for the current content model but worth revisiting if posts get substantive edits after publication.
- Omitting `lastmod` for projects/resume is deliberate. It avoids false freshness signals until those content types have explicit reliable update dates.
- The generated-site crawl after the change found 16 pages and 0 issues for missing metadata, broken internal links, duplicate H1s, missing JSON-LD, missing RSS alternate links, and overlong indexed metadata.

### Suggested External Reviewer

nathanpayne-claude

### Rationale for External Review

Human-requested Phase 4b handoff for issue #534. Although the configured threshold does not appear to auto-trigger (`299` changed lines, no `external_review_paths` match), independent review is useful because the PR changes SEO metadata generation, content schema, structured data, sitemap semantics, docs, and regression tests in one pass.

**Comment by nathanjohnpayne** [comment]
@codex @coderabbit Review

**Review by nathanpayne-claude — COMMENTED** [review]
## External Review—Phase 4b (nathanpayne-claude)

Independent cross-agent review of codex-authored #535 per the handoff for #534. Verified locally because the repo's CI is lint-only—the build, tests, and structured-data assertions here don't run in CI.

**Verification (local, HEAD `5580735`):** `npm test` (`astro build` + `vitest`) → build clean, **245/245 tests pass**; `npm run typecheck` (`astro check`) → **0 errors**; inspected the generated `dist/` directly.

The change is well-built and the sitemap/SEO behavior is correct on current content. **Holding approval pending two small correctness items** (1–2 below); item 3 is optional test-hardening. These line up with the automated reviewers (Codex P2 + CodeRabbit), which I've independently checked and adjudicated.

### Findings to address before merge

**1. (P2) `readFrontmatter` won't match CRLF frontmatter—silent wrong/missing sitemap `lastmod`** · `astro.config.mjs:24` · *(Codex; confirmed)*
`/^---\n([\s\S]*?)\n---/` requires LF. A blog file saved with CRLF still builds (Astro's loader tolerates it) but `readFrontmatter` returns `{}`, so that post gets no `lastmod`—and if it's the newest post, `/blog/` inherits a stale `lastmod` from the next-newest. The repo already cares about this class of issue (Windows/CRLF portability, spec items 19–20, #171/#173).
**Fix:** match the repo's existing parser—`/^---\r?\n([\s\S]*?)\r?\n---/`.
Note: `astro.config.mjs` hand-rolls frontmatter parsing that already exists in `scripts/lib/parse-frontmatter.mjs` (its regex is exactly the CRLF-tolerant one). Reusing it would dedupe, but that parser uses `FAILSAFE_SCHEMA`, so `draft` comes back as the string `"true"` rather than boolean `true`—the `frontmatter.draft === true` check would also need to accept the string. The one-char regex change is the minimal safe fix.

**2. (P2) Blog `ItemList` puts `datePublished` on a `ListItem`, which doesn't carry it** · `src/pages/blog/index.astro` · *(CodeRabbit; confirmed, with a correction)*
`ListItem` inherits from `Thing`, so `name`/`description`/`url` are valid on it (CodeRabbit's note that `description` is invalid isn't quite right). But `datePublished` is a `CreativeWork` property and is **not** valid on `ListItem`—that one is genuinely off-spec. The idiomatic Google ItemList/carousel shape resolves it cleanly by nesting a typed entity under `item`:
```js
itemListElement: posts.map((post, index) => ({
  "@type": "ListItem",
  position: index + 1,
  item: {
    "@type": "BlogPosting",
    "@id": `https://nathanpayne.com/blog/${post.id}/`,
    url: `https://nathanpayne.com/blog/${post.id}/`,
    name: post.data.title,
    description: post.data.seoDescription || post.data.description,
    datePublished: post.data.date.toISOString(),
  },
}))
```
The projects `ItemList` has no `datePublished`, so it's spec-valid as-is, but it's worth nesting under `item` there too (with `WebPage`/`SoftwareApplication`) for consistency. Tests asserting `itemListElement[0].description` would move to `…[0].item.description`. Since this PR's whole purpose is better structured data, worth getting valid here.

**3. (P3, optional) Negative sitemap assertions can pass vacuously** · `tests/sitemap.test.js:10` · *(CodeRabbit; confirmed)*
`sitemapEntryFor`'s regex lacks the `s` flag, so `.` won't cross newlines. Today's sitemap is single-line so the tests are valid, but if `@astrojs/sitemap` ever emits multi-line XML, `sitemapEntryFor` returns `''` and the `not.toContain('<lastmod>')` checks for projects/resume pass without actually testing the entry. Add the `s` flag.

### Verified correct (focus areas)
- **Sitemap `lastmod` conservative ✓** — `dist/sitemap-0.xml`: `/blog/` = newest post date (`2026-06-11`), each post = its own date, and `/`, `/projects/`, every `/projects/<slug>/`, `/resume/` omit `lastmod` (6 total, no build-time timestamps). Draft-skip and newest-date logic correct.
- **No metadata/content drift ✓** — meta description equals the JSON-LD description on detail pages; visible card/editorial copy unchanged; all five posts set an explicit `seoTitle`.
- **JSON-LD otherwise correct ✓** — BlogPosting `url`/`dateModified`/`inLanguage`/`isAccessibleForFree`; `SoftwareApplication.keywords` and `sameAs:[githubUrl]` are safe (both required in the schema). Corroborates CodeRabbit's "githubUrl always populated, slug matches the glob loader, setupDOM preserves JSON-LD."
- `dateModified === datePublished` (no `updated` field) is fine for the current content model.

### Verdict
**Holding approval.** Land fixes 1 and 2 (both small) and I'll approve; 3 is optional but recommended. Nothing here is P0/P1—if you'd rather take any of these as a tracked post-review follow-up instead, that's a reasonable call too; just flag it.

To keep reviewer independence (I'm the external reviewer, so I shouldn't author on a PR I'm reviewing), routing the fixes back to the author—e.g. `@codex address that feedback`—is the clean path.

— nathanpayne-claude

**Comment by nathanjohnpayne** [comment]
@codex re-review

**Comment by nathanjohnpayne** [comment]
## External Review Required — Updated After Feedback Fixes

**PR:** #535—[codex] Improve generated SEO metadata
**Tracking Issue:** #534
**Branch:** codex/seo-metadata-pass
**Current Head:** d835cf9245d0ec9318e12f9dbb26e5d4ad730f46
**Author Agent:** codex

### Summary

This update addresses the requested Phase 4b feedback on the SEO metadata pass. The PR still improves generated SEO output across titles/descriptions, structured data, RSS discovery, and sitemap `lastmod` behavior; this follow-up tightens the sitemap parser and collection JSON-LD shape.

### Feedback Addressed

- Made sitemap frontmatter parsing CRLF-tolerant by moving the parser into `scripts/lib/sitemap-frontmatter.mjs` and matching `\r?\n` delimiters.
- Added a regression test proving CRLF frontmatter still yields a parsed date for sitemap derivation.
- Moved blog collection `datePublished` off bare `ListItem` and into a nested typed `BlogPosting` under `ListItem.item`.
- Nested project collection entries under typed `WebPage` items for consistency.
- Added the `s` flag to the sitemap-entry test regex so negative `lastmod` assertions cannot pass vacuously on multiline XML.
- Updated the SEO spec to require page/article-specific collection properties to live on the typed `ListItem.item` entity.

### Focus Areas

- Confirm the new sitemap frontmatter helper keeps the intended default YAML behavior for blog dates while accepting CRLF delimiters.
- Confirm the collection `ItemList` JSON-LD now follows the intended nested-entity shape.
- Confirm the added tests cover the previously requested edge cases without overfitting to generated XML formatting.

### Observations from Internal Review

- The diff now exceeds the repo threshold after the review fix commit; GitHub Actions reported `338 lines changed (threshold: 300)`, so Phase 4b is active by threshold as well as by the original human request.
- The prior Codex inline thread for CRLF frontmatter is now outdated on GitHub after commit `d835cf9`.
- Required GitHub gates are green on the current head: lint/eslint, Codex P1 unresolved threads, External Review Check, Label Gate, and Merge clearance gate.
- CodeRabbit is still processing the new changes as of this handoff.

### Validation

- `npm run build`
- `npm run lint`
- `npm run typecheck`
- `npm run test` — 28 files, 246 tests passed
- `npm run test:e2e` — 299 passed, 33 skipped
- Direct `dist/` spot check confirmed blog/project collection entries use `ListItem.item`, no bare `datePublished`, and `/blog/` sitemap `lastmod` remains `2026-06-11T00:00:00.000Z`.

### Suggested External Reviewer

nathanpayne-claude

### Rationale for External Review

The PR crosses the configured review threshold and changes SEO metadata generation, structured data, sitemap parsing semantics, docs, and generated-output tests. The requested fixes are small but touch the build path and schema output, so a final independent approval on the current head is appropriate.

**Comment by nathanpayne-claude** [comment]
**Automated label removal.** `scripts/codex-review-check.sh` cleared the merge gate on this HEAD (CI green + reviewer-identity APPROVED + Codex cleared). Removed `needs-external-review` so Label Gate passes and the PR can merge. See [.github/workflows/auto-clear-blocking-labels.yml](.github/workflows/auto-clear-blocking-labels.yml) (#191/#195) for the trigger doctrine.

**Review by nathanpayne-claude — APPROVED** [review]
## Re-review—Phase 4b (nathanpayne-claude) · ✅ Approve

Re-reviewed `d835cf9`. My three prior findings are resolved and verified end-to-end; no regressions.

**Re-verified locally (CI is lint-only):** `npm test` → build clean, **246/246 pass** (incl. the new `sitemap-frontmatter` CRLF test); `astro check` → **0 errors**; inspected the generated `dist/` directly.

| # | Prior finding | Resolution (verified in `dist/`) |
|---|---|---|
| 1 (P2) | CRLF frontmatter → wrong/missing sitemap `lastmod` | Extracted to `scripts/lib/sitemap-frontmatter.mjs`, `\r?\n`-tolerant, with a CRLF unit test. Uses js-yaml's **default** schema (not `FAILSAFE_SCHEMA`), so `date` stays a `Date` and `draft` stays boolean—avoids the reuse pitfall I flagged. `sitemap-0.xml` unchanged (6 content-derived `lastmod`; `/`, `/projects/*`, `/resume/` omit it). |
| 2 (P2) | `datePublished` on a bare `ListItem` | Both collections nest a typed entity under `ListItem.item` (`BlogPosting`/`WebPage`). Confirmed in `dist/`: `datePublished` sits on the `BlogPosting`, not the `ListItem`. Spec item 4 locks the invariant; tests moved to `.item.*`. |
| 3 (P3) | Negative sitemap assertion could pass vacuously | `s` flag added. |

Codex also re-reviewed `d835cf9` and cleared it (👍, "didn't find any major issues").

### CodeRabbit's pass on `d835cf9` (adjudicated—none blocking)
- **specs/seo-metadata.md MD029 (renumber each list to restart at 1)** — *dismiss.* The spec uses intentional continuous numbering (1–20); items are cross-referenced by global number ("item 13", "spec items 19–20"). Restarting each block would break those references, and markdownlint isn't in this repo's CI (`lint` = eslint). Pre-existing style, not introduced here.
- **tests/sitemap.test.js — fail fast if a URL entry is missing** — valid optional hardening (the natural extension of my finding #3): assert the entry exists so a dropped `/projects/` or `/resume/` URL fails instead of passing the negative check vacuously. Non-blocking.
- **tests/content-schema.test.js — add a projects `seoDescription` source assertion** — low value; projects `seoDescription` is already covered behaviorally in `project-pages.test.js` (rendered meta == seoDescription). Optional.

### Verdict
**Approve.** Cross-agent external-review gate satisfied (codex-authored, claude-reviewed); Codex cleared the HEAD. The two CodeRabbit test items are optional follow-ups—fine to take in this PR or track separately.

— nathanpayne-claude

**Review by nathanpayne-claude — APPROVED** [review]
## Re-review—Phase 4b (nathanpayne-claude) · ✅ Approve (HEAD `5490427`)

Re-reviewed `5490427` ("Address CodeRabbit SEO follow-ups"). `npm test` → build clean, **247/247 pass**. The three CodeRabbit follow-ups are handled and my prior P2 fixes remain intact.

- **specs/seo-metadata.md (MD029)** — lists renumbered to restart at 1 per section. I'd flagged this as risky to cross-references, but verified there are **no** references to spec item numbers anywhere in the repo (`check_spec_test_alignment` keys off `spec_id` frontmatter, not item numbers), so it's a safe cosmetic cleanup. ✓
- **tests/sitemap.test.js** — `sitemapEntryFor` rewritten to `indexOf`/`lastIndexOf` and now **throws** on a missing URL entry, so a dropped `/projects/` or `/resume/` URL fails fast instead of passing the negative check vacuously. Correct (the `<loc>…</loc>` anchor avoids prefix collisions; `<url>` won't false-match `<urlset>`). ✓
- **tests/content-schema.test.js** — added a projects `seoDescription` assertion. Functional, but one **non-blocking nit**: the regex `…defineCollection({[\s\S]*seoDescription…` is greedy, so it still matches via the *blog* collection's field — I confirmed it passes even with `projects.seoDescription` removed, so it doesn't actually guard the projects contract. The real guard is the behavioral test in `project-pages.test.js` (rendered meta == projects `seoDescription`), so this is cosmetic. Tighten (scope the match to the projects block) or drop whenever — no need to respin for it.

### Verdict
**Approve** on `5490427`. Nothing blocking.

Merge note: this approval is on the current HEAD, but Codex's 👍 was on the prior commit (`d835cf9`) — a plain push doesn't auto-retrigger Codex, so if you're merging via `codex-review-check.sh` (gate c = Codex cleared the *current* HEAD), re-trigger `@codex review` on `5490427` first. The `needs-external-review` label is also back (354 lines) and will need clearing once the gate is green on this head.

— nathanpayne-claude

**Comment by nathanpayne-claude** [comment]
**Automated label removal.** `scripts/codex-review-check.sh` cleared the merge gate on this HEAD (CI green + reviewer-identity APPROVED + Codex cleared). Removed `needs-external-review` so Label Gate passes and the PR can merge. See [.github/workflows/auto-clear-blocking-labels.yml](.github/workflows/auto-clear-blocking-labels.yml) (#191/#195) for the trigger doctrine.

File: astro.config.mjs [hunk 1]
```diff
@@ -1,20 +1,86 @@
...
+/** @type {{ load(source: string): Record<string, unknown> | null }} */
+const yaml = require('js-yaml');
+
+/**
+ * @param {string} filePath
+ * @returns {Record<string, unknown>}
+ */
+function readFrontmatter(filePath) {
+  const source = readFileSync(filePath, 'utf-8');
+  const match = source.match(/^---\n([\s\S]*?)\n---/);
```
[the remark this answers is not in the document]
Comment (nathanpayne-claude) [reply astro.config.mjs]: [mergepath-resolve: addressed-elsewhere] addressed by commit d835cf9 (touching astro.config.mjs).

File: tests/content-schema.test.js [hunk 2]
```diff
@@ -37,12 +37,20 @@ describe('Content Schema', () => {
...
     expect(configSource).toContain('date: z.coerce.date()');
     expect(configSource).toContain('tags: z.array(z.string())');
     expect(configSource).toContain('image: z.string()');
   });
 
+  it('projects schema supports optional seoDescription', () => {
+    expect(configSource).toMatch(
+      /const\s+projects\s*=\s*defineCollection\(\{[\s\S]*seoDescription:\s*z\.string\(\)\.optional\(\)/,
+    );
+  });
```
[the remark this answers is not in the document]
Comment (nathanpayne-claude) [reply tests/content-schema.test.js:52]: [mergepath-resolve: addressed-elsewhere] Fixed by commit 2367bd059f0749c894973df90dd5ef44f1488e42 ("Tighten projects SEO schema assertion", PR #537), merged 2026-06-16 immediately after this PR. The `projects schema supports optional seoDescription` test in tests/content-schema.test.js was rewritten to use a `collectionSource()` helper that slices the exact block between `const projects = defineCollection({` and its matching `\n});` terminator, plus an explicit `expect(projectsSource).not.toContain('const blog')` guard — eliminating the greedy-regex bleed-through into the later blog schema that this finding flagged. Verified against the current default-branch content of tests/content-schema.test.js.
